### PR TITLE
Milestone 4

### PR DESCRIPTION
**Pull request by GurevichVarvara**
So, this is 4 milestone and I have couple of questions.

For that moment I've created routing which is partly implemented within my flask-project-side (could I call it back-end in the future?) and another part of implementation is inside of login-register_form_confirmation.js file: there I redirect user to main page after successful authentification. And here my questions start.

First question preface: I look at the login-procces like that: first of all we should make 'first-step-validation', and by that I mean usernames length validation (and whether the password and its confirmation are identical if we talk about register form), and if this validation was successful we send entry to server wich is performing validation on the presonal users data level.
First question: I think we should do 'first-step-validation' inside js-file, because it's not necessery to send whole response to the server and recive another response with error-message (it takes time actualy) just to know that usernames length not valid. But on the other hand if one day I'd like to change this affordable length I would prefer to change it inside my python-files on the server-side, not in js on the client-side. What would you recomend me about that?

Second question: Is it allowed to perform routing within python-files on the server-side? I know how to do it via js but at the same time I'm not sure it would be good in the architectural way. (Everything related to routing and controling login-status to prevent access to pages which is not allowed to be visited by not logged in users you could find in app.py file in the "TODO" project-folder.)

Third question preface: For now my project is not created for huge amount of users, that is why I dodn't use sql-bd for storing data, insted I have property "users" inside my Database-class where I have methods to work with user's data, info about that users I store in pickle-format cause it's the most convinient way to store serialized objects. Inside that users I have its username, password hash (and list of habits and todos in the future). I don't want to have db on my vps, but if my storage method is not acceptable I will come up with a solution for this situation.

Fourth qestsion: I use Fetch API, and I saw that you prevent us from using any API. Fetch is also one of them? or I could use it because I understand how it works?

That's all for now. I'm not sure is that amount of work enough for that moment, if not then tell me what to add and I'll do it.

I changed Firebase to vps, so new link is: http://62.171.133.174:8000/

### CODE REVIEW

**Comment by gmltA** [comment]
> First question

A good practice here is to handle validation both on client and server side. Client-only validation is not secure, because you can freely edit client-side scripts. The only secure place to validate data is on the server side. Nevertheless you are absolutely right about sending to the server pieces of data which are 100% not correct (like short passwords). You can implement those checks in JS code to display early errors to the user.

> Second question

The goal here is to perform routing client-side with asynchronous request to server only to obtain data (possibly with pieces of markup, if you want), but not the whole new page.


> if my storage method is not acceptable

I'm only interested in front-end side of your project. You can utilise any degree of anti-patterns on back-end side as long as it helps you prototype web-app faster. Of course, it is applicable only in terms of this project.

> Fourth qestsion: I use Fetch API, and I saw that you prevent us from using any API. Fetch is also one of them? or I could use it because I understand how it works?

That's absolutely wrong. I **do** encourage usage of all native platform API's such as fetch, notification API, etc. The only thing I am mostly against is a unnecessarily included libraries (just as adding jQuery to perform AJAX requests).

File: TODO/static/JS/login-register_form_confirmation.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+
+function check_username_length(username) {
+    var result = false;
+    if (username.length < 15) {
+        result = true;
+    }
+
+    return result;
```
Comment (gmltA) [inline TODO/static/JS/login-register_form_confirmation.js:21]:
```suggestion
    return username.length < 15;
```

Am I missing something?

Comment (GurevichVarvara) [reply TODO/static/JS/login-register_form_confirmation.js]: Fair enough)

**Comment by GurevichVarvara** [comment]
I've fixed that couple of things. Are there any other points that I should pay attention to?

File: TODO/static/JS/login-register_form_confirmation.js [hunk 2]
```diff
@@ -12,14 +12,8 @@ window.onload=function() {
 }
 
 
-function check_username_length(username) {
-    var result = false;
-    if (username.length < 15) {
-        result = true;
-    }
+var check_username_length = (username) => { return username.length < 15; };
```
Comment (gmltA) [inline TODO/static/JS/login-register_form_confirmation.js:15]: `var` is redundant. Stick to `const` or `let`.

**Pull request merged**